### PR TITLE
First pull request of repository to review 8 initial files of preprocessing and linear models.

### PR DESCRIPTION
**Pull request by ivichadriana**
The following pull request contains 8 new files (7 Jupyter Notebooks and one Python script). These are for data preprocessing for linear regression analysis (in the preprocessing folder), linear regression analysis (in the exploratory experiments folder), and one file to prepare files for CIBERSORTx. 
The files were very big to do in one single pull request, so some files were previously added to this repository beforehand. All of these files are considered "new" and need review I know it's a lot; this is the first pull request of the repository. 

The linear_models_* files are very similar, with only slight changes in order to filter/not filter the data. The files in preprocessing folder are also very similar with only slight changes in cell-type predictions.
 
The files are:
- cibersort_prep.ipynb (file to prep data for CIBERSORTx).
- linear_models_bulk.ipynb (liner reg. with bulked-up data).
- linear_models_diffcelltype_filtered.ipynb (liner reg. with one SN cell-type predicting all others in SC).
- linear_models_samecelltype_filtered.ipynb (liner reg. with one SN cell-type predicting the same cell-type in SC).
- translator_fcn.py (script with functions used).
- filtered_scaled_scsn_mousekidney.ipynb (preprocessing and QC with filtering and scaling).
- notfiltered_notscaled_scsn_mousekidney.ipynb (preprocessing and QC without filtering or scaling).
- notfiltered_scaled_scsn_mousekidney.ipynb (preprocessing and QC without filtering but yes scaling).

### CODE REVIEW

File: exploratory_experiments/__pycache__/translator_fcn.cpython-39.pyc
Comment (nrosed) [inline exploratory_experiments/__pycache__/translator_fcn.cpython-39.pyc:1]: you probably dont wants this folder or file in the repo

File: exploratory_experiments/translator_fcn.py [hunk 1]
```diff
@@ -28,7 +28,6 @@
 import pickle
 from pathlib import Path
 
-
 def linear_model_cell_gene(adata, obs_var, obs_cat_X, obs_cat_Y, num_test, num_train):
     #based on https://scikit-learn.org/stable/auto_examples/linear_model/plot_ols.html
     #this fcn withholds genes to train and test
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:33]: on line 40, you are selecting a specific subset of the matrix, then taking the mean. It isn't clear here what obs_var, obs_cat_X, etc, should be -- I am guessing they are cell_types, but it may be useful to either use more descriptive variable names or add more comments describing what the variables are. I am guilty of not doing this, but the best practice is to add a docstring for each method. An example is here: https://pandas.pydata.org/docs/development/contributing_docstring.html#:~:text=A%20Python%20docstring%20is%20a,html)%20documentation%20automatically%20from%20docstrings.

File: exploratory_experiments/translator_fcn.py [hunk 2]
```diff
@@ -80,7 +79,6 @@ def linear_model_cell_gene(adata, obs_var, obs_cat_X, obs_cat_Y, num_test, num_t
 
     return(regr, x_train, y_train, x_test, y_test, y_pred)
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:80]: instead of showing the plot, since this is an internal method, would you instead want to return the plt object for the user of the method can choose to plot it or not?

File: exploratory_experiments/translator_fcn.py [hunk 3]
```diff
@@ -80,7 +79,6 @@ def linear_model_cell_gene(adata, obs_var, obs_cat_X, obs_cat_Y, num_test, num_t
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:79]: also on line 47-51, you split your data into test and train, but they are not shuffled. I would recommend that you shuffle your data before splitting it. 

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

File: exploratory_experiments/translator_fcn.py [hunk 4]
```diff
@@ -99,8 +97,8 @@ def linear_model_cellnum(adata, obs_var, obs_cat_X, obs_cat_Y, pc_test):
     idx1 = int(len(y)*pc_test)
     y_test= np.mean(y[0:idx1,:], axis=0) 
     y_test = y_test[:,np.newaxis]
-    idx2 = int(len(x)*pc_train)
-    y_train = np.mean(x[idx2:-1,:], axis=0)
+    idx2 = int(len(y)*pc_train)
+    y_train = np.mean(y[idx2:-1,:], axis=0)
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:101]: I would suggest to use https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
for test train split

File: exploratory_experiments/translator_fcn.py [hunk 5]
```diff
@@ -99,8 +97,8 @@ def linear_model_cellnum(adata, obs_var, obs_cat_X, obs_cat_Y, pc_test):
     idx1 = int(len(y)*pc_test)
     y_test= np.mean(y[0:idx1,:], axis=0) 
     y_test = y_test[:,np.newaxis]
-    idx2 = int(len(x)*pc_train)
-    y_train = np.mean(x[idx2:-1,:], axis=0)
+    idx2 = int(len(y)*pc_train)
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:100]: its not really clear to me how this method differs from linear_model_cell_gene, I think a docstring would help here.

File: exploratory_experiments/translator_fcn.py [hunk 6]
```diff
@@ -131,7 +129,6 @@ def linear_model_cellnum(adata, obs_var, obs_cat_X, obs_cat_Y, pc_test):
 
     return(regr, x_train, y_train, x_test, y_test, y_pred)
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:130]: I recommend returning the plot object when using internal methods.

File: exploratory_experiments/translator_fcn.py [hunk 7]
```diff
@@ -173,8 +170,8 @@ def linear_model_diffcell_gene(adata, obs_var, x, ct1, ct2, pc_test):
     idx1 = int(len(yte)*pc_test)
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:170]: I recommend adding a docstring -- its not clear what ct1 and ct2 mean here, but I am guessing they are a required type of index variable. Might also be useful to add a check to make sure that the fstring (f"single_nucleus_{ct1}") is present in `obs` -- otherwise you will return back an empty object and the user may not know it.

File: exploratory_experiments/translator_fcn.py [hunk 8]
```diff
@@ -173,8 +170,8 @@ def linear_model_diffcell_gene(adata, obs_var, x, ct1, ct2, pc_test):
...
     y_test= yte[0:idx1]
     y_test = y_test[:,np.newaxis]
-    idx2 = int(len(xtr)*pc_train)
-    y_train = xtr[idx2:-1]
+    idx2 = int(len(ytr)*pc_train)
+    y_train = ytr[idx2:-1]
     y_train = y_train[:,np.newaxis]
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:175]: I would suggest to use more descriptive variable names -- specifically idx1 and idx2, what are these indexing?

File: exploratory_experiments/translator_fcn.py [hunk 9]
```diff
@@ -271,6 +268,86 @@ def linear_model_diffcellnum(adata, obs_var, x, ct1, ct2, pc_test):
 
     return(regr, x_train, y_train, x_test, y_test, y_pred)    
 
+def linear_model_bulked(adata1, adata2, pc_test, by_what, x,y):
+
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:272]: what is adata1 and adata2? It's typically not best practice to have variable names that are not descriptive and only differ by a number. It may make sense to you, but other users will likely find it confusing to read your code -- even when you have a docstring added. For readability, it's better to have longer variable names that make your code easier to read than short variable names.

File: exploratory_experiments/translator_fcn.py [hunk 10]
```diff
@@ -271,6 +268,86 @@ def linear_model_diffcellnum(adata, obs_var, x, ct1, ct2, pc_test):
...
+        if idx2 < 1:
+            print("Can't separate by cells, doing by genes")
+            idx2 = 1
+            by_what = "genes"
+        y_test = adata2.X[0:idx2,:]
+        y_train= adata2.X[idx2:-1,:]
+        print(f"ytrain is {y_train.shape}")
+
+    #taking average across all cells of the same celltype, keeping all genes or cells
+    y_train= np.mean(y_train, axis=0) #SC
```
Comment (nrosed) [inline exploratory_experiments/translator_fcn.py:310]: IF for this method you always are training on sc and testing on SN, instead of having the comment on the side #SC and #SN, could you make y_test variable name include "sc", so like sc_test or y_test_sc?

File: exploratory_experiments/translator_fcn.py (hunk 8 above)
Comment (nrosed) [reply exploratory_experiments/translator_fcn.py:175]: also xtr to x_train and xte to x_test

File: preprocessing/filtered_innerouter_snsc_mousekidney.ipynb [hunk 11]
```diff
@@ -210,90 +210,6 @@
     "bulk_adata = bulk_adata.transpose()"
```
Comment (nrosed) [inline preprocessing/filtered_innerouter_snsc_mousekidney.ipynb:210]: <pre class="ql-syntax" spellcheck="false">Line #3.    plt.title('MSE of Bulk Expr. vs. Average Expr. in SN and SC (filtered)')
</pre><p>can you show this after taking the log of the MSE? Its hard to see the distribution when there are so many large outliers.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/greenelab/rnaseq_translator/pull/1/discussion/#comment-1181821401'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='128'/>

Comment (nrosed) [inline preprocessing/filtered_innerouter_snsc_mousekidney.ipynb:210]: <pre class="ql-syntax" spellcheck="false">Line #2.    pca_X1 = pca_10.fit_transform(sn_adata.X) #sort of like the training dimensions
</pre><p>instead of X1 and X2 I would suggest to use more descriptive variable names</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/greenelab/rnaseq_translator/pull/1/discussion/#comment-1181821405'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='130'/>

Comment (nrosed) [inline preprocessing/filtered_innerouter_snsc_mousekidney.ipynb:210]: <pre class="ql-syntax" spellcheck="false">Line #2.    plt.xlabel('Num of Components')
</pre><p>This plot looks very strange -- is your sn.adata.X matrix transposed? Its just confusing that so little variation is explained. How many samples do you have? Can you plot the PCA plot to visually see why so little variation is explained? If it is this little variation, the plot should look almost perfectly spherical.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/greenelab/rnaseq_translator/pull/1/discussion/#comment-1181821408'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='133'/>

Comment (nrosed) [inline preprocessing/filtered_innerouter_snsc_mousekidney.ipynb:210]: <pre class="ql-syntax" spellcheck="false">Line #1.    pca_095 = PCA(n_components=0.95)
</pre><p>hmmm how is the number of components less than 1? what does this mean?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/greenelab/rnaseq_translator/pull/1/discussion/#comment-1181821409'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='136'/>

**Review by nrosed — COMMENTED** [review]
Overall I think it looks good. Main suggestions are

1. intros to the notebooks
2. more descriptive variable names
3. check either the linear regression fits or the normalization of the data. Something looks a bit strange there, OR I may be misunderstanding the plots/coefficients/MSE that are reported.

**Review by nrosed — APPROVED** [review]
Overall I think it looks good. Main suggestions are

1. intros to the notebooks
2. more descriptive variable names
3. check either the linear regression fits or the normalization of the data. Something looks a bit strange there, OR I may be misunderstanding the plots/coefficients/MSE that are reported.
